control/utils.py

```
@@ -662,3 +665,117 @@ def __str__(self):
             f"ip v4: {','.join(self.ipv4_addresses)}\n"
             f"ip v6: {','.join(self.ipv6_addresses)}\n"
         )
+
+
+class DsaUtils:
+    def __init__(self, logger):
+        self.logger = logger
+
+    def _build_dsa_config(self):
+        pci_id = ("0x8086", "0x0b25")
+        base_path = Path("/sys/bus/pci/devices")
+        json_list = []
+        count = 0
+
+        for dev_path in base_path.iterdir():
+            vendor_file = dev_path / "vendor"
+            device_file = dev_path / "device"
+
+            if not vendor_file.exists() or not device_file.exists():
+                continue
+
+            vendor = vendor_file.read_text().strip()
+            device = device_file.read_text().strip()
+
+            if (vendor, device) == pci_id:
```

according to:
[https://doc.dpdk.org/guides/dmadevs/idxd.html](https://doc.dpdk.org/guides/dmadevs/idxd.html) in 5.3:
_Intel® DSA devices can use the IDXD kernel driver or DPDK-supported drivers, such as vfio-pci. Both are supported by the IDXD PMD.
..To use Intel® DSA devices in DPDK multi-process applications, the devices should be bound to the vfio-pci driver. Multi-process is not supported when using the kernel IDXD driver._

IMO need to add something like:
```python3
def _verify_driver_binding(self, dev_path):
    driver_file = dev_path / "driver"
    if not driver_file.exists():
        self.logger.warning(f"DSA device {dev_path.name} has no driver")
        return False
    
    driver_name = driver_file.resolve().name
    if driver_name != "idxd":
        self.logger.warning(f"DSA device bound to {driver_name}, need idxd for kernel mode")
        return False
    
    return True
```
The driver check ensures your kernel-mode DSA configuration will actually work, preventing silent failures where devices are detected but unusable

---

the user has not the ability to chose the driver, user get complete built-in node with software + hardware